### ISSUE TITLE
Feature to see what the current postgres → oracle activity is?

### ISSUE THREAD

**Issue by alexitheodore** [laurenz/oracle_fdw#396]
From what I've been able to tell this feature doesn't exist and I cannot find any older mentions of such a request in closed issues, so here it is:

I am dealing with a setup where I am syncing between Postgres and Oracle often but the Oracle server is very slow and basically appears as a black box to me from the Postgres side (_which is all I have access to_). It would be really handy to have some visibility into what open threads there are. 

**TLDR: is/can there be an `oracle_fdw_stat_activity()` view?**

**Comment by laurenz** [comment]
Since oracle_fdw execution is tied into the query executor, I think that anything like that would require changes to PostgreSQL to show the current executor state.

I am wondering: what kind of information do you expect? The back-seat child's "Will we be there soon?" is probably out of reach.

**Comment by laurenz** [comment]
Another idea: you have client access to the Oracle database.
So you could query `V$SESSION_LONGOPS` to check Oracle's prognosis.
You could even create a foreign table to access this information via oracle_fdw.

**Comment by alexitheodore** [comment]
I'm purposefully being open-ended to my specifications and requirements because I don't know what the limitations are and I also don't have anything terribly specific in mind. Whatever can be provided would be appreciated. The general idea of what I am asking for is an insight into what requests have been queued and what the status of those requests are. 

By status, I don't expect anything like "are we there yet" but something like a list of open requests, including the contents of the request (the postgres and corresponding oracle query), the execution start time and the last communication (such as if a cancel message has been sent).

I assume this hasn't been done or possibly even considered because the expectation might be that requests are so short-lived that a list of their statuses would basically always be empty. But that isn't necessarily true. In my situation I wait around for sometimes hours for a query to finish - all the while I wonder what is happening.

Sure, in an ideal circumstance I could probably find out *something* like this information by going to Oracle directly - but that defeats the point of a fdw function that is there either for a true answer, or convenience or for the cases where that is either not an option or doesn't quite answer the need. For example:

1. In my case, my oracle user permissions don't allow me to view a list of running queries (don't ask me how that works).
2. Even if I did have a list I would suspect that it might get complicated parsing out the list that I get from Oracle from the list that the fdw extension is truly active on. Why should I be reinventing the wheel when it should be better to actually ask the fdw itself what it is doing?

Thank you for your consideration!

**Comment by laurenz** [comment]
I understand the wish, but features like that so far only exist for certain &ldquo;utility satements&rdquo; like `VACUUM` in PostgreSQL, not for queries in general. A feature like that would involve PostgreSQL core and can hardly be implemented in the foreign data wrapper alone.

It is not a case of re-inventing the wheel, but of inventing the wheel, because no such functionality currently exists. It might be a quite useful wheel, because you are certainly not the only person who wonders how long running queries are doing.

I am surprised at your reluctance to use `V$SESSION_LONGOPS`; perhaps you misunderstood. What I suggest is that you use `CREATE FOREIGN TABLE` to create a PostgreSQL foreign table to examine that Oracle view. If you don't have permission to see these data, then oracle_fdw couldn't do so either, so that's something you'll have to take up with your Oracle DBAs.

Viewing the plans of running PostgreSQL queries is not possible with core PostgreSQL, but you could have a look at the [pg_show_plans](https://github.com/cybertec-postgresql/pg_show_plans) extension.

I don't think I can build anything into oracle_fdw to make this as nice and simple as you would wish, but I think that my suggestions can go a long way towards helping you with your problem.

**Comment by alexitheodore** [comment]
Thank you for the response and the clarifications. I am fine accepting the answer that you have given, even if it is not the one I hoped for. I appreciate the thought and effort you put into responding to me.

As for the V$SESSION_LONGOPS; suggestion - the truth is that I am working as a sub-contractor to a contractor for a large company. I am working with stuff that I do not have any control over or any contacts with whom I can address permissions. Obviously that is my problem and not necessarily a common scenario.

**Comment by laurenz** [comment]
Sorry to hear that. Often it is just a question of asking and talking to people.
They want you to get a job done, so they should be interested in your problems.
